core/src/main/java/org/apache/iceberg/rest/RestCatalog.java

```diff
@@ -0,0 +1,324 @@
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.rest.http.ErrorHandlers;
+import org.apache.iceberg.rest.http.HttpClient;
+import org.apache.iceberg.rest.http.RequestResponseSerializers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RestCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces, Configurable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RestCatalog.class);
+
+  private String catalogName;
+  private Map<String, String> properties;
+  private ObjectMapper mapper;
+  private String baseUrl;
+  private Configuration hadoopConf;
+  private FileIO fileIO;
+
+  // TODO - Refactor to use interface, which doesn't currently have POST etc embedded into it.
+  private HttpClient restClient;
+
+  @Override
+  public void initialize(String name, Map<String, String> props) {
+    this.catalogName = name;
+    this.properties = props;
+
+    // TODO - Possibly authenticate with the server initially and then have the server return some of this information
+    Preconditions.checkNotNull(
+        properties.getOrDefault("baseUrl", null),
+        "Cannot initialize the RestCatalog as the baseUrl is a required parameter.");
+
+    this.baseUrl = properties.get("baseUrl");
+    this.fileIO = initializeFileIO(properties);
+
+    this.mapper = new ObjectMapper();
+    RequestResponseSerializers.registerAll(mapper);
+
+    // TODO - We can possibly handle multiple warehouses via one RestCatalog to reuse the connection pool
+    //        and for cross database calls if users need to authenticate with each.
+    this.restClient = HttpClient.builder()
+        .baseUrl(String.format("%s/warehouse/%s", baseUrl, catalogName))
+        .mapper(mapper)
+        .defaultErrorHandler(ErrorHandlers.tableErrorHandler())
+        .build();
+  }
+
+  // TODO - Pass in ObjectMapper too when testing.
+  @VisibleForTesting
+  public void initialize(
+      String name,
+      Map<String, String> props,
+      HttpClient httpClient) {
+    this.catalogName = name;
+    this.properties = props;
+
+    // TODO - Possibly authenticate with the server initially and then have the server return some of this information
+    Preconditions.checkNotNull(
+        properties.getOrDefault("baseUrl", null),
+        "Cannot initialize the RestCatalog as the baseUrl is a required parameter.");
+
+    this.baseUrl = properties.get("baseUrl");
+    this.fileIO = initializeFileIO(properties);
+
+    this.mapper = new ObjectMapper();
```

Comment:
I think it is recommended to use object mapper as a singleton. We can reuse `JsonUtil.mapper()` and add util methods around that to register modules.

And actually why not directly register the serdes in that mapper directly, instead of only registering it when initializing the REST catalog? I think it does not hurt to do that even for people not using REST catalog.

Reply:
That would certainly make it simpler. And I don't see the harm in registering the serdes directly, which would make a lot of things simpler. Good suggestion. 👍